src/main/java/org/openstreetmap/atlas/checks/validation/areas/SpikyBuildingCheck.java

```diff
@@ -0,0 +1,417 @@
+    }
+
+    /**
+     * For an point defined by the two surrounding segments, return the angle and location of that
+     * point if that point is not part of a curve, and the angle between the two segments is less
+     * than headingThreshold.
+     * 
+     * @param beforeAngle
+     *            the segment directly before the point in question
+     * @param afterAngle
+     *            the segment directly after the point in question
+     * @param curvedLocations
+     *            the locations of all curved segments in the polygon
+     * @return an empty optional if the point is part of a curve, or if the angle is greater than or
+     *         equal to headingThreshold. Otherwise, a tuple containing the location of the point
+     *         and the angle between beforeAnge and afterAngle
+     */
+    private Optional<Tuple<Angle, Location>> getSkinnyAngleLocation(final Segment beforeAngle,
+            final Segment afterAngle, final Set<Location> curvedLocations)
+    {
+        if (!curvedLocations.contains(afterAngle.end())
+                && !curvedLocations.contains(beforeAngle.start()))
+        {
+            final Angle difference = this.getDifferenceBetween(beforeAngle, afterAngle.reversed(),
+                    Angle.MAXIMUM);
+            if (difference.isLessThan(headingThreshold))
+            {
+                return Optional.of(Tuple.createTuple(difference, afterAngle.start()));
+            }
+        }
+        return Optional.empty();
+    }
+
+    /**
+     * Gets the difference in headings between firstSegment and secondSegment, returning
+     * defaultAngle if either segments are a point.
+     * 
+     * @param firstSegment
+     *            the first segment to compare
+     * @param secondSegment
+     *            the second segment to compare
+     * @param defaultAngle
+     *            the default value to return
+     * @return the difference between firstSegment.heading() and secondSegment.heading() if neither
+     *         segment is a single point (same start and end nodes), or defaultAngle if either
+     *         segment is a single point
+     */
+    private Angle getDifferenceBetween(final Segment firstSegment, final Segment secondSegment,
+            final Angle defaultAngle)
+    {
+        return firstSegment.heading()
+                .flatMap(first -> secondSegment.heading().map(first::difference))
+                .orElse(defaultAngle);
+    }
+
+    @Override
+    protected Optional<CheckFlag> flag(final AtlasObject object)
+    {
+        final List<Tuple<Angle, Location>> allSkinnyAngles = this.getPolygons(object)
+                .map(this::getSkinnyAngleLocations)
+                .filter(angleLocations -> !angleLocations.isEmpty()).flatMap(Collection::stream)
+                .collect(Collectors.toList());
+        if (!allSkinnyAngles.isEmpty())
+        {
+            final Set<AtlasObject> objectsToFlag;
+            if (object instanceof Area)
```

Comment:
This could be a ternary statement inside the createFlag call below. That way you don't even need to use a singleton Set.

Reply:
I just tried this, but the following block of code: 
```
            return Optional.of(this.createFlag(object instanceof Area ? object : ((Relation) object).flatten(),
                    this.getLocalizedInstruction(0, headingThreshold.toString(),
                            allSkinnyAngles.stream().map(Tuple::getFirst).map(Angle::toString)
                                    .collect(Collectors.joining(", "))),
                    allSkinnyAngles.stream().map(Tuple::getSecond).collect(Collectors.toList())));
```
gives the following error:
```
error: no suitable method found for createFlag(object ins[...]ten(),String,List<Location>)
            return Optional.of(this.createFlag(object instanceof Area ? object : ((Relation) object).flatten(),
                                   ^
    method BaseCheck.createFlag(AtlasObject,String,List<Location>) is not applicable
      (argument mismatch; bad type in conditional expression
          Set<AtlasObject> cannot be converted to AtlasObject)
    method BaseCheck.createFlag(Set<AtlasObject>,String,List<Location>) is not applicable
      (argument mismatch; bad type in conditional expression
          AtlasObject cannot be converted to Set<AtlasObject>)
```
Basically, the java compiler isn't smart enough to call one over the other depending on the type returned by a ternary statement. Unless perhaps I misunderstood what you said.

I revised this anyways, opting for not using the singleton Set over using the ternary statement, but if you like it better the other way I can switch it!

Reply:
Interesting, I thought I had used that before, but I guess not. Your revision looks good.